keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,103 @@
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by using an admission controller to check if the associated StorageClasses have `allowVolumeExpansion = true`.
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
+
+While updating a pod, the StatefulSet controller will update a referenced persistent volume claim object if its storage request in the associated `volumeClaimTemplate` has been increased.
+
+Not all volumes support online control-plane expansion so this design aims to support both online and offline control-plane volume expansion as follows.
+
+Prerequisite: The StatefulSet controller must be able to determine if a volume supports online expansion.
+
+To accomplish this a new method `RequiresOfflineResize` will be added to the `ExpandableVolumePlugin` interface. Through this it will possible to determine which in-tree volume plugins support online expansion. 
+
+For volumes provisioned through the CSI, an optional boolean field `SupportsOnlineExpansion` will be added into the `CSIDriverSpec` field of the `CSIDriver` object. The `CSIDriver` object stores information about a CSI volume driver deployed on the cluster. If the `CSIDriver` object or the `SupportsOnlineExpansion` field is not registered for a particular volume provisioned through the CSI, the statefulset controller will assume the volume does not support online expansion. Adding this functionality mandates an API change for the mentioned objects.
+
+If a volume is deemed to support online expansion it will be expanded before the pod is terminated for the update. The StatefulSet controller will wait for the file system resize to complete on all such volumes before terminating the associated pod.
+
+Otherwise, the volume will be expanded offline after the pod is terminated and before it is recreated.
+
+This design minimizes the time a StatefulSet pod is unavailable if all volumes support online expansion while still supporting volumes that can only be expanded ofline. 
+
+Note that the Statefulset controller does not support in place updates for any pod modifications. Therefore, even if all referenced volumes support online expansion the associated pods will still need to be deleted and immediately recreated before completing the update. Support for in place updates may be introduced at a later point in time.
```

Comment:
You mean, we have to do a full update of all pods?  You're implying the update strategy would run?  Or that a new type of update would exist?  Generally we wouldn't add a new type of update, we'd just enforce the existing update mechanism and force the controller to consider resize of claim template to be the same as updating parts of the pod spec.

Reply:
I'm not sure what exactly you mean by a full update, but the update to the pvc will occur when the pod is being updated. So we are not adding a new type of update rather just following the selected update strategy. The pvc update code is agnostic to the update strategy. This [PR ](https://github.com/kubernetes/kubernetes/pull/71384/files) gives an idea of where the pvc will be updated. Ofcourse a bunch of changes need to be made to get it on par with this proposal.

The in-place update part is just to document the consideration we had that maybe we'd skip the pod delete step if the only change made to the statefulset was updating the pvc. We decided not to go through with this since none of the workload controllers support in place modifications and it would be beyond the scope of this enhancement.